feat: adds callback hooks for connection state changes, surfaces connection state

jacoscaz opened the pull request:
This PR addresses #48 by adding callback hooks to the `Client` class which may be used by consumers to subscribe to changes in the state of the MQTT connection. Additionally, this PR surfaces `Context#connectionState` to consumers through the `Client` class, so that it may be polled to understand the state of the connection without waiting for a state change.

Whenever possible, I find that it is always better to keep event-based code and async/await code separate and let downstream consumers bridge the gap, if needed. Therefore, I've opted to delay execution of the callbacks via `queueMicrotask()` and kept them as simple functions rather than model them as asynchronous ones.

jacoscaz commented on client/client.ts:
```diff
@@ -164,7 +178,9 @@ export class Client {
         this.connectPacket.clean = false;
         this.ctx.close();
       } catch (err) {
+        // TODO: can we replace this with an assert that ensures err to be instanceof Error?
```
@seriousme question for you: if we can safely assume `err` to be an instance of `Error`, can we simply cast it into an `Error` at compile-time or assert it to be an `Error` at run time?

seriousme replied:
I checked, we currently can't assume `err` to be an instance:
https://github.com/seriousme/opifex/blob/0430e16fb68a94f942f5078d806344bed216c0d3/client/client.ts#L131

And similarly:
https://github.com/seriousme/opifex/blob/0430e16fb68a94f942f5078d806344bed216c0d3/node/tcpClient.ts#L99

and:
https://github.com/seriousme/opifex/blob/0430e16fb68a94f942f5078d806344bed216c0d3/deno/tcpClient.ts#L73

But we could change all 3 of  them to produce an Error object. We just need to make sure then that in `bin/mqtt.js` we create a nice error message.

jacoscaz replied:
That seems to be already the case: 

https://github.com/seriousme/opifex/blob/0430e16fb68a94f942f5078d806344bed216c0d3/bin/mqtt.ts#L211

I've wrapped the strings into Error objects in those three lines and added assertions to make sure caught values are instanceof errors.

jacoscaz commented:
Should be ready to review. Not sure as to why the `deno fmt --check` step is failing, it appears to suggest the same thing it's complaining about.

jacoscaz commented:
Thank you for the lovely library and the super-extra-quick review turnaround!

> Btw: do you plan to do more improvements on short term or should we publish a new release now?

I think I'm ok for the time being. I've fixed everything that I had stumbled in while integrating the library into my project (which I'll publish at some point in the near future) and I've added what I needed to add. Can't promise I won't open another PR next week but, for the time being, I don't see anything urgent.

jacoscaz commented:
Awesome, thank you @seriousme !
